src/melange_json.mli

```diff
@@ -0,0 +1,434 @@
+    |}
+  )
+]}
+*)
+
+type t = Js.Json.t
+(** The type of a JSON data structure *)
+
+type json = t
+(** Defined for convenience. *)
+
+val to_string : json -> string
+(** JSON can be encoded as a string. *)
+
+type exn +=
+  | Of_string_error of string
+        (** The exception raised when parsing JSON error occurs *)
+
+val of_string : string -> json
+(** JSON can be parsed from a string. Raises {Of_string_error}. *)
+
+type 'a to_json = 'a -> json
+(** Describe how to encode a value into JSON. *)
+
+val to_json : json to_json
+(** JSON can be encoded as JSON, trivially. *)
+
+(** The type of a error which occurs during decoding JSON values. *)
+type of_json_error = Json_error of string | Unexpected_variant of string
+
+val of_json_error_to_string : of_json_error -> string
+val of_json_error : ?depth:int -> ?width:int -> json:json -> string -> 'a
+val of_json_msg_error : string -> 'a
+
+type exn +=
+  | Of_json_error of of_json_error
+        (** The exception raised when a decoding error occurs *)
+
+type 'a of_json = json -> 'a
+(** Describes how to decode a value out of JSON. *)
+
+val of_json : json of_json
+(** JSON can be decoded from JSON, trivially. *)
+
+module Of_json : sig
+  (** Provides a set of low level combinator primitives to decode
+      Js.Json.t data structures A decoder combinator will return the
+      decoded value if successful, or raise a [Of_json_error] exception if
+      unsuccessful, where the string argument contains the error message.
+      Decoders are designed to be combined to produce more complex
+      decoders that can decode arbitrary data structures, though the
+      emphasis for this library is for it to be {i possible} to decode any
+      given data structure, not necessarily for it to be {i convenient}.
+      For convenience you should look towards opinionated third-party
+      libraries. *)
+
+  val string : string of_json
+  val char : char of_json
+  val bool : bool of_json
+  val int : int of_json
+  val int64 : int64 of_json
+  val float : float of_json
+  val unit : unit of_json
+  val array : 'a of_json -> 'a array of_json
+  val list : 'a of_json -> 'a list of_json
+  val option : 'a of_json -> 'a option of_json
+  val tuple2 : 'a of_json -> 'b of_json -> ('a * 'b) of_json
+
+  val tuple3 :
+    'a of_json -> 'b of_json -> 'c of_json -> ('a * 'b * 'c) of_json
+
+  val tuple4 :
+    'a of_json ->
+    'b of_json ->
+    'c of_json ->
+    'd of_json ->
+    ('a * 'b * 'c * 'd) of_json
+
+  val result : 'a of_json -> 'b of_json -> ('a, 'b) result of_json
+
+  (** Auxiliary combinators *)
+
+  val field : string -> 'a of_json -> 'a of_json
+  val at : string list -> 'a of_json -> 'a of_json
+  val one_of : 'a of_json list -> 'a of_json
+  val either : 'a of_json -> 'a of_json -> 'a of_json
+  val try_or_none : 'a of_json -> 'a option of_json
+  val try_of_default : 'a -> 'a of_json -> 'a of_json
+  val map : ('a -> 'b) -> 'a of_json -> 'b of_json
+
+  (** Some JS specific combinators. *)
```

Comment:
How can we have Js specific combinators in the united runtime ?

Reply:
The united runtime is not in the direction you're suggesting. The unification is for:
- ppx runtime
- library runtime

not melange/ocaml or js/native unification.